src/tty.rs

```diff
@@ -201,92 +178,73 @@ fn get_pw_entry(buf: &mut [i8; 1024]) -> Passwd {
     }
 }
 
-/// Exec a shell
-fn execsh(config: &Config) -> ! {
+/// Create a new tty and return a handle to interact with it.
+pub fn new<T: ToWinsize>(config: &Config, size: T) -> Pty {
+    let win = size.to_winsize();
     let mut buf = [0; 1024];
     let pw = get_pw_entry(&mut buf);
 
-    let shell = match config.shell() {
-        Some(shell) => match shell.to_str() {
-            Some(shell) => shell,
-            None => die!("Invalid shell value")
-        },
-        None => pw.shell
-    };
-
-    // setup environment
-    env::set_var("LOGNAME", pw.name);
-    env::set_var("USER", pw.name);
-    env::set_var("SHELL", shell);
-    env::set_var("HOME", pw.dir);
-    env::set_var("TERM", "xterm-256color"); // sigh
-
-    unsafe {
-        libc::signal(libc::SIGCHLD, libc::SIG_DFL);
-        libc::signal(libc::SIGHUP, libc::SIG_DFL);
-        libc::signal(libc::SIGINT, libc::SIG_DFL);
-        libc::signal(libc::SIGQUIT, libc::SIG_DFL);
-        libc::signal(libc::SIGTERM, libc::SIG_DFL);
-        libc::signal(libc::SIGALRM, libc::SIG_DFL);
-    }
-
-    // pw.shell is null terminated
-    let shell = unsafe { CStr::from_ptr(shell.as_ptr() as *const _) };
-
-    let argv = [shell.as_ptr(), ptr::null()];
-
-    let res = unsafe {
-        libc::execvp(shell.as_ptr(), argv.as_ptr())
-    };
+    let (master, slave) = openpty(win.ws_row as _, win.ws_col as _);
+    let mut shell = OsString::from(pw.shell);
```

Comment:
These manual conversions to `OsStr`/`OsString` shouldn't be necessary. `process::Command::new()` takes `AsRef<OsStr>` which is implemented for `&str`, `String`, `Path`, and `PathBuf` (all the types we care about).

`&str`/`Path` should be preferred since they don't require copies _and_ because they implement `AsRef<OsStr>`.